File: 473-matchsticks-to-square/473-matchsticks-to-square.cpp
class Solution {
public:
    bool makeSquareHelper(vector<int> &matchSticks,int n,vector<int> &squares,int target){
        if(n == matchSticks.size()){
            for(int i=1;i<4;i++){
                if(squares[i-1]!=squares[i]) return false;
            }
            return true;
        }
                
        for(int i=0;i<4;i++){
            if(squares[i]+matchSticks[n]>target) continue;
            squares[i] += matchSticks[n];
            if(makeSquareHelper(matchSticks,n+1,squares,target)) return true;
            squares[i] -= matchSticks[n];
        }
        
        return false;
    }    
    
    bool makesquare(vector<int>& matchSticks) {
        int n = matchSticks.size();
        vector<int> squares(4,0);
        int sum=0;
        sum = accumulate(matchSticks.begin(),matchSticks.end(),0);
        if(sum%4 !=0) return false;
        sort(matchSticks.rbegin(),matchSticks.rend());
        return makeSquareHelper(matchSticks,0,squares,sum/4);
    }
};

File: 473-matchsticks-to-square/473-matchsticks-to-square.java
class Solution {
    public boolean makesquare(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }

        int sum = 0;
        for (int num : nums) {
            sum += num;
        }

        if (sum % 4 != 0) {
            return false;
        }

        Arrays.sort(nums);
        reverse(nums);
        int[] squareEdge = new int[4];
        return recMakeSquare(0, squareEdge, nums, sum / 4);
    }

    public boolean recMakeSquare(int index, int[] squareEdge, int[] nums, int target) {
        if (index == nums.length) {
            return squareEdge[0] == squareEdge[1] && squareEdge[1] == squareEdge[2];
        }

        for (int i = 0; i < 4; i++) {
            if (squareEdge[i] + nums[index] > target) {
                continue;
            }

            squareEdge[i] += nums[index];
            if (recMakeSquare(index + 1, squareEdge, nums, target)) {
                return true;
            }
            squareEdge[i] -= nums[index];
        }
        return false;
    }

    public void reverse(int[] nums) {
        for (int i = 0; i < nums.length / 2; i++) {
            int temp = nums[i];
            nums[i] = nums[nums.length - 1 - i];
            nums[nums.length - 1 - i] = temp;
        }
    }
}